tokenclient/tokenFlows.go

```diff
@@ -0,0 +1,196 @@
+// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Cloud Security Client Go contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+package tokenclient
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"github.com/sap/cloud-security-client-go/auth"
+	"github.com/sap/cloud-security-client-go/env"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+type Options struct {
+	HTTPClient *http.Client // TODO Default: basic http.Client with a timeout of 15 seconds
+	TLSConfig  *tls.Config  // TODO Default:
+}
+
+// RequestOptions allows to configure the token request
+type RequestOptions struct {
+	// Context carries the request context like the deadline or other values that should be shared across API boundaries.
+	Context context.Context
+	// Request parameters that shall be overwritten or added to the payload
+	Params map[string]string
+}
+
+type TokenFlows struct {
+	identity *env.Identity
+	options  Options
+	tokenURI string
+}
+
+type ClientError struct {
+	msg string
+	err error
+}
+
+func (e *ClientError) Error() string {
+	if e.err == nil {
+		return e.msg
+	}
+	return fmt.Sprintf("%s: %s", e.msg, e.err.Error())
+}
+
+type tokenResponse struct {
+	Token string `json:"access_token"`
+}
+
+const (
+	tokenEndpoint              string = "/oauth2/token" //nolint:gosec
+	grantTypeParameter         string = "grant_type"
+	grantTypeClientCredentials string = "client_credentials"
+	clientIDParameter          string = "client_id"
+	clientSecretParameter      string = "client_secret"
+)
+
+// NewTokenFlows initializes token flows
+//
+// identity provides credentials and url to authenticate client with identity service
+// options specifies rest client and its tls config, both can be overwritten.
+// Note: Setup of default tls config is not supported for windows os. Module crypto/x509 supports SystemCertPool with go 1.18 (https://go-review.googlesource.com/c/go/+/353589/)
+func NewTokenFlows(identity *env.Identity, options Options) (*TokenFlows, *ClientError) {
```

Comment:
What's the reason behind letting the consumer choose tls config and client? Would it be possible to just specify one (or none) of these?

Reply:
Good question! yes, it should be possible to use DefaultTLSConfig , and then apps only need to provide HttpClient => I've added it in the type doc as well.

What do you think, does it make sense or do you think, applications will need to configure the cert pool by themselves in almost all cases?

Reply:
for now i've simplified it and moved it into httpclient package, so that reuse is possible

Reply:
pls check again